### PR TITLE
[Enabler][1377]mvs_raw_support_gdg_gds_special_character

### LINKED ISSUE #1377: [Enabler] [zos_mvs_raw] GDG/GDS and special character support
# Enabler description

New MVSDataset class
-----------------------

There is a new MVSDataset class introduced in #1386 that represents a data set, when used, as part of initialization it escapes and resolves the gds relative name if any.

This is a sample on how to initialize the new class, many of these params can just be None.
```python
MVSDataSet(
            name = params.get("name"),
            record_format = params.get("record_format", None),
            volumes = params.get("volumes", None),
            data_set_type = params.get("type", None),
            block_size = params.get("block_size", None),
            record_length = params.get("record_length", None),
            space_primary = params.get("space_primary", None),
            space_secondary = params.get("space_secondary", None),
            space_type = params.get("space_type", None),
            directory_blocks = params.get("directory_blocks", None),
            key_length = params.get("key_length", None),
            key_offset = params.get("key_offset", None),
            sms_storage_class = params.get("sms_storage_class", None),
            sms_data_class = params.get("sms_data_class", None),
            sms_management_class = params.get("sms_management_class", None),
        )
```

GDG/GDS Support
-------------------

zos_mvs_raw takes data set names in different options as dd inputs, we should support GDG relative name whenever a data set name is accepted.

```yaml
- name: Take a set of data sets and write them to an archive.
  zos_mvs_raw:
    program_name: adrdssu
    auth: yes
    dds:
      - dd_data_set:
          dd_name: archive
          data_set_name: MYHLQ.ARCHIVE
          disposition: old
      - dd_data_set:
          dd_name: sysin
          data_set_name: ZOAUSER.GDG(0)
          disposition: shr
      - dd_dummy:
          dd_name: sysprint

```

Internally, in the case of a data set name being a GDS relative name we should resolve those names into absolute GDS names when the generation number is 0 or negative.

Following functions will be implemented in #1386 
```python
# Sample of soon to be implemented gdg function
if DataSet.is_gds_relative_name(name):
    src = DataSet.resolve_gds_absolute_name(src)
```

Symbols support
-----------------

**We need to verify if symbols work natively here or if we need to implement this**

Add symbol support. 

data set name with symbol

`ZOAUSER.TE$T`

That way, the modules will have to retrieve the new escaped named and the original user provided name inside the dictionary like:

```python
name = DataSet.escape_data_set_name(name)
```

Escaping special characters could be done via a new function inside the BetterArgParser.
These are some of the special characters:
- \#
- @
- $
- \-

For this task completion we need:
- [ ] Do the development work
- [ ] Update the docs to talk about GDG and GDS handling in the module.
- [ ] Update the docs with at least 1 example on how to use with GDSs.
- [ ] Create a new test case for GDSs.
- [ ] Create a new test case for symbols.

### Ansible module

zos_mvs_raw

### PR DESCRIPTION
**Pull request by AndreMarcel99**
##### SUMMARY

Fixes #1377 

##### ISSUE TYPE

- Enabler Pull Request

##### COMPONENT NAME
Add function resolve_data_set_names for dd_data_set

##### ADDITIONAL INFORMATION
When get key values of the dd's at the moment to check dd_dataset check for the data_set_name (required parameter) by first check if is a relative positive if the disposition is new will work properly other case will check if the data set exist and disposition is new it will fail other case return the absolute name.
Captura de pantalla 2024-05-30 a la(s) 9 59 25 a m
Captura de pantalla 2024-05-30 a la(s) 9 59 33 a m

### CODE REVIEW

**Comment by AndreMarcel99** [comment]
Changes on` plugins/module_utils/data_set.py` are from block and lineinfile work, that's why I didn't test

**Review by richp405 — APPROVED** [review]

File: plugins/modules/zos_mvs_raw.py [hunk 1]
```diff
@@ -2589,6 +2627,42 @@ def set_extra_attributes_in_dd(dd, tmphlq, key):
...
+      -------
+      str
+          The absolute name of dataset or relative positive if disposition is new.
+    """
+    if data_set.DataSet.is_gds_relative_name(dataset):
+        if data_set.DataSet.is_gds_positive_relative_name(dataset):
+            if disposition and disposition == "new":
+                return dataset
+            else:
+                raise ("To generate a new gds as {0} required disposition new".format(dataset))
```
Comment (fernandofloresg) [inline plugins/modules/zos_mvs_raw.py:2649]:
```suggestion
                raise ("To generate a new GDS as {0} disposition 'new' is required.".format(dataset))
```

File: plugins/modules/zos_mvs_raw.py [hunk 2]
```diff
@@ -2589,6 +2627,42 @@ def set_extra_attributes_in_dd(dd, tmphlq, key):
...
+        else:
+            data = data_set.MVSDataSet(
+                name=dataset
+            )
+            src = data.name
+            if data.is_gds_active:
+                if disposition and disposition == "new":
+                    raise ("Gds {0} already create, incorrect parameters {1} and {2}".format(src, "disposition", "data_set_name"))
```
Comment (fernandofloresg) [inline plugins/modules/zos_mvs_raw.py:2657]:
```suggestion
                    raise ("GDS {0} already created, use a positive generation when disposition  '{1}' is used.".format(src, "disposition", "data_set_name"))
```

**Review by fernandofloresg — CHANGES_REQUESTED** [review]
Just left a few suggestions

File: plugins/modules/zos_mvs_raw.py [hunk 3]
```diff
@@ -2589,6 +2627,42 @@ def set_extra_attributes_in_dd(dd, tmphlq, key):
...
+            else:
+                raise ("To generate a new GDS as {0} disposition 'new' is required.".format(dataset))
+        else:
+            data = data_set.MVSDataSet(
+                name=dataset
+            )
+            src = data.name
+            if data.is_gds_active:
+                if disposition and disposition == "new":
+                    raise ("Gds {0} already create, incorrect parameters for disposition and data_set_name".format(src))
```
Comment (fernandofloresg) [inline plugins/modules/zos_mvs_raw.py:2657]:
```suggestion
                    raise ("GDS {0} already created, incorrect parameters for disposition and data_set_name".format(src))
```

**Review by fernandofloresg — APPROVED** [review]

**Review by ketankelkar — APPROVED** [review]

**Pull request merged**